sign/bls/bls.go

```diff
@@ -0,0 +1,414 @@
+		n, err := io.ReadFull(rd, OKM)
+		if n != len(OKM) || err != nil {
+			return nil, err
+		}
+
+		ss.SetBytes(OKM)
+
+		if ss.IsZero() == 1 {
+			digest := sha256.Sum256(salt)
+			salt = digest[:]
+		} else {
+			return &PrivateKey[K]{key: ss, pub: nil}, nil
+		}
+	}
+
+	return nil, ErrKeyGen
+}
+
+// Sign computes a signature of a message using a key (defined in
+// G1 or G1).
+func Sign[K KeyGroup](k *PrivateKey[K], msg []byte) Signature {
+	if !k.Validate() {
+		panic(ErrInvalidKey)
+	}
+
+	switch (interface{})(k).(type) {
+	case *PrivateKey[G1]:
+		var Q GG.G2
+		Q.Hash(msg, []byte(dstG2))
+		Q.ScalarMult(&k.key, &Q)
+		return Q.BytesCompressed()
+	case *PrivateKey[G2]:
+		var Q GG.G1
+		Q.Hash(msg, []byte(dstG1))
+		Q.ScalarMult(&k.key, &Q)
+		return Q.BytesCompressed()
+	default:
+		panic(ErrInvalid)
+	}
+}
+
+// Verify returns true if the signature of a message is valid for the
+// corresponding public key.
+func Verify[K KeyGroup](pub *PublicKey[K], msg []byte, sig Signature) bool {
+	var (
+		a, b interface {
+			setBytes([]byte) error
+			hash([]byte)
+		}
+		listG1 [2]*GG.G1
+		listG2 [2]*GG.G2
+	)
+
+	switch (interface{})(pub).(type) {
+	case *PublicKey[G1]:
+		aa, bb := new(G2), new(G2)
+		a, b = aa, bb
+		k := (interface{})(pub.key).(G1)
+		listG1[0], listG1[1] = &k.g, GG.G1Generator()
+		listG2[0], listG2[1] = &aa.g, &bb.g
+	case *PublicKey[G2]:
+		aa, bb := new(G1), new(G1)
+		a, b = aa, bb
+		k := (interface{})(pub.key).(G2)
+		listG2[0], listG2[1] = &k.g, GG.G2Generator()
+		listG1[0], listG1[1] = &aa.g, &bb.g
+	default:
+		panic(ErrInvalid)
+	}
+
+	err := b.setBytes(sig)
+	if err != nil {
+		return false
+	}
+	if !pub.Validate() {
+		return false
+	}
+	a.hash(msg)
+
+	res := GG.ProdPairFrac(listG1[:], listG2[:], []int{1, -1})
+	return res.IsIdentity()
+}
+
+// Aggregate produces a unified signature given a list of signatures.
+// To specify the group of keys pass either G1{} or G2{} as the first
+// parameter.
+func Aggregate[K KeyGroup](k K, sigs []Signature) (Signature, error) {
+	if len(sigs) == 0 {
+		return nil, ErrAggregate
+	}
+
+	switch (interface{})(k).(type) {
+	case G1:
+		var P, Q GG.G2
+		P.SetIdentity()
+		for _, sig := range sigs {
+			if err := Q.SetBytes(sig); err != nil {
+				return nil, err
+			}
+			P.Add(&P, &Q)
+		}
+		return P.BytesCompressed(), nil
+
+	case G2:
+		var P, Q GG.G1
+		P.SetIdentity()
+		for _, sig := range sigs {
+			if err := Q.SetBytes(sig); err != nil {
+				return nil, err
+			}
+			P.Add(&P, &Q)
+		}
+		return P.BytesCompressed(), nil
+
+	default:
+		panic(ErrInvalid)
+	}
+}
+
+// VerifyAggregate returns true if the aggregated signature is valid for
+// the list of messages and public keys provided. The slices must have
+// equal size and have at least one element.
+func VerifyAggregate[K KeyGroup](pubs []*PublicKey[K], msgs [][]byte, aggSig Signature) bool {
+	if len(pubs) != len(msgs) || len(pubs) == 0 || len(msgs) == 0 {
+		return false
+	}
+
+	n := len(pubs)
```

Comment:
I think you forgot the public key validation.

Reply:
added now